### DISCUSSION TITLE
Improve performance by using fixed point multiplication to apply volume

### DISCUSSION THREAD

**Discussion by LinusHeu** [pschatzmann/arduino-audio-tools#926] [Ideas]
https://github.com/pschatzmann/arduino-audio-tools/blob/af3e2c50eb54fb978176f35799f1dd1a6220e720/src/AudioTools/VolumeStream.h#L279-L288

(factorForChannel() returns a float.)

Currently, volume is applied by using floats. 

https://forums.raspberrypi.com/viewtopic.php?t=308794#p1848263
According to this benchmark, RP2040 can do around 1,5 million floating point operations and 5,7 million fixed point operations per second. 

There could be a significant performance increase by switching to fixed point multiplication here. I think the user should set the volume as a float (for convenience) and this library should convert it to fixed point.
I plan to further look into this and maybe implement it. Unless you want to do it. Or is there a reason why fixed point math should not be used? If fixed point leads to worse performance on MCUs with an FPU, there could be a #DEFINE to switch between float or fixed point. 

As comparison, ESP8266Audio uses fixed point math:
https://github.com/earlephilhower/ESP8266Audio/blob/0abcf71012f6128d52a6bcd155ed1404d6cc6dcd/src/AudioOutput.h#L67C1-L72C6

**Comment by pschatzmann** [comment]
It was getting the most precise results and I was never running into performance problems.
So I left it with float.

**Reply by LinusHeu** [reply]
The reason I'm looking into this is, that playing a wav from SD works great on RP2040. But when I added volume control, it sounded like the performance wasn't good enough.
Of course, I could also run the SD card faster or overclock the pico. And I haven't completely confirmed whether it's actually a performance issue or something else. I just did a quick "wav from SD + volume control" test a few weeks ago, and am just now getting back into it.

**Comment by pschatzmann** [comment]
It would be interesting to do some performance comparisions on different processors. I think I remember that floating points were faster on the ESP32.

This might be interesting https://github.com/pschatzmann/arduino-audio-tools/wiki/Filters
The RP2040 seems to be pretty bad with floating points.

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/pull/945
